formula/src/main/java/com/instacart/formula/Effect.kt

```diff
@@ -0,0 +1,47 @@
+package com.instacart.formula
+
+/**
+ * Effect is a function returned within [Transition.Result] which will be executed by
+ * [FormulaRuntime] usually after [Formula] state changes are applied. This function is
+ * the place to call listeners, log events, trigger network requests or database writes,
+ * and etc. Instead of executing effects within [Transition.toResult], deferring execution
+ * allows us to ensure that [Formula] is always in the correct state in case effects
+ * trigger a new event.
+ */
+data class Effect(
+    val type: Type,
+    val executable: () -> Unit,
+) {
+    constructor(executable: () -> Unit): this(Main, executable)
+
+    /**
+     * An effect type which defines the execution timing and threading of the effect. Take
+     * a look at [Unconfined], [Main], [Background].
+     */
+    sealed class Type
+
+    /**
+     * Unconfined effect type indicates that the effect can run on any thread and will be
+     * executed after all state changes are processed.
+     */
+    data object Unconfined : Type()
+
+    /**
+     * Immediate effect type indicates that the effect can run on any thread and can be executed
+     * immediately before the state changes are processed and formula is re-evaluated. This
+     * is a useful optimization for inlining multiple state changes as part of a single
+     * re-evaluation.
```

Comment:
Not sure I'm following here. Isn't it in Formula's contract that effects are guaranteed to be executed *after* state changes application?

Reply:
It's for special cases where you might not want that behavior. For example, if a child formula updates its state and then notifies the parent via callback to also update. Currently, this would have two full paths to evaluate child state change, perform the side-effect, and then evaluate the parent state. With the `Immediate` type, we would update child state, then the parent state, and only then, do a single re-evaluation.

Reply:
Note: this will be implemented in the future (just added placeholder for the type to communicate the purpose)